# Load selected models of all datasets

In [3]:
import pickle

In [4]:
# load selected models for regression datasets
best_model_series = pickle.load(open('best_model_series.pkl','rb'))

In [5]:
# load selected models for classification datasets
best_Cmodel_series = pickle.load(open('best_Cmodel_series.pkl','rb'))

# Install Libraries

In [ ]:
pip install PyTDC

In [ ]:
pip install rdkit-pypi

In [8]:
import numpy as np
import pandas as pd

from rdkit import Chem
from rdkit.Chem import Descriptors, Lipinski

# Creating global features extractor function

In [9]:
def molecular_descriptors(table):

  descriptors = pd.DataFrame()

  mol = [Chem.MolFromSmiles(drug) for drug in table.Drug]

  # Exact molecular weight of the molecule
  Nilavo = []
  Nilavo.append([Descriptors.ExactMolWt(i) for i in mol])
  descriptors['Exact_MW'] = Nilavo[0]

  # FpDensityMorgan1
  Nilavo = []
  Nilavo.append([Descriptors.FpDensityMorgan1(i) for i in mol])
  descriptors['FpDensityMorgan1'] = Nilavo[0]

  # FpDensityMorgan2
  Nilavo = []
  Nilavo.append([Descriptors.FpDensityMorgan2(i) for i in mol])
  descriptors['FpDensityMorgan2'] = Nilavo[0]

  # FpDensityMorgan3
  Nilavo = []
  Nilavo.append([Descriptors.FpDensityMorgan3(i) for i in mol])
  descriptors['FpDensityMorgan3'] = Nilavo[0]

  # Average molecular weight of the molecule ignoring hydrogens
  Nilavo = []
  Nilavo.append([Descriptors.HeavyAtomMolWt(i) for i in mol])
  descriptors['HeavyAtomMolWt'] = Nilavo[0]

  ###
  ### MaxAbsPartialCharge ###
  Nilavo = []
  Nilavo.append([Descriptors.MaxAbsPartialCharge(i) for i in mol])
  descriptors['MaxAbsPartialCharge'] = Nilavo[0]

  ###
  ### MaxPartialCharge ###
  Nilavo = []
  Nilavo.append([Descriptors.MaxPartialCharge(i) for i in mol])
  descriptors['MaxPartialCharge'] = Nilavo[0]

  ###
  ### MinAbsPartialCharge ###
  Nilavo = []
  Nilavo.append([Descriptors.MinAbsPartialCharge(i) for i in mol])
  descriptors['MinAbsPartialCharge'] = Nilavo[0]

  ###
  ### MinPartialCharge ###
  Nilavo = []
  Nilavo.append([Descriptors.MinPartialCharge(i) for i in mol])
  descriptors['MinPartialCharge'] = Nilavo[0]

  # Average molecular weight of the molecule
  Nilavo = []
  Nilavo.append([Descriptors.MolWt(i) for i in mol])
  descriptors['MolWt'] = Nilavo[0]

  # Number of radical electrons of the molecule
  Nilavo = []
  Nilavo.append([Descriptors.NumRadicalElectrons(i) for i in mol])
  descriptors['NumRadicalElectrons'] = Nilavo[0]

  # Number of valence electrons of the molecule
  Nilavo = []
  Nilavo.append([Descriptors.NumValenceElectrons(i) for i in mol])
  descriptors['NumValenceElectrons'] = Nilavo[0]

  # Log of partition coefficient
  Nilavo = []
  Nilavo.append([Descriptors.MolLogP(i) for i in mol])
  descriptors['Partition_Coefficient'] = Nilavo[0]


  ### Lipinski Descriptors ###
  # Fraction of C atoms that are SP3 hybridized
  Nilavo = []
  Nilavo.append([Lipinski.FractionCSP3(i) for i in mol])
  descriptors['FractionCSP3'] = Nilavo[0]

  # Number of heavy atoms a molecule
  Nilavo = []
  Nilavo.append([Lipinski.HeavyAtomCount(i) for i in mol])
  descriptors['Heavy_atoms'] = Nilavo[0]

  # Number of NHs or OHs
  Nilavo = []
  Nilavo.append([Lipinski.NHOHCount(i) for i in mol])
  descriptors['NHs/OHs'] = Nilavo[0]

  # Number of Nitrogens and Oxygens
  Nilavo = []
  Nilavo.append([Lipinski.NOCount(i) for i in mol])
  descriptors['N&O'] = Nilavo[0]

  # Number of aliphatic (containing at least one non-aromatic bond) carbocycles for a molecule
  Nilavo = []
  Nilavo.append([Lipinski.NumAliphaticCarbocycles(i) for i in mol])
  descriptors['Aliphatic_carbocycles'] = Nilavo[0]

  # Number of aliphatic (containing at least one non-aromatic bond) heterocycles for a molecule
  Nilavo = []
  Nilavo.append([Lipinski.NumAliphaticHeterocycles(i) for i in mol])
  descriptors['Aliphatic_heterocycles'] = Nilavo[0]

  # Number of aliphatic (containing at least one non-aromatic bond) rings for a molecule
  Nilavo = []
  Nilavo.append([Lipinski.NumAliphaticRings(i) for i in mol])
  descriptors['Aliphatic_rings'] = Nilavo[0]

  # Nmber of aromatic carbocycles for a molecule
  Nilavo = []
  Nilavo.append([Lipinski.NumAromaticCarbocycles(i) for i in mol])
  descriptors['Aromatic_carbocycles'] = Nilavo[0]

  # Number of aromatic heterocycles for a molecule
  Nilavo = []
  Nilavo.append([Lipinski.NumAromaticHeterocycles(i) for i in mol])
  descriptors['Aromatic_heterocycles'] = Nilavo[0]

  # Number of aromatic rings for a molecule
  Nilavo = []
  Nilavo.append([Lipinski.NumAromaticRings(i) for i in mol])
  descriptors['Aromatic_rings'] = Nilavo[0]

  # Number of Hydrogen Bond Acceptors
  Nilavo = []
  Nilavo.append([Lipinski.NumHAcceptors(i) for i in mol])
  descriptors['HAcceptors'] = Nilavo[0]

  # Number of Hydrogen Bond Donors
  Nilavo = []
  Nilavo.append([Lipinski.NumHDonors(i) for i in mol])
  descriptors['HDonors'] = Nilavo[0]

  # Number of Heteroatoms
  Nilavo = []
  Nilavo.append([Lipinski.NumHeteroatoms(i) for i in mol])
  descriptors['Heteroatoms'] = Nilavo[0]

  # Number of Rotatable Bonds
  Nilavo = []
  Nilavo.append([Lipinski.NumRotatableBonds(i) for i in mol])
  descriptors['Rotatable_Bonds'] = Nilavo[0]

  # Number of saturated carbocycles for a molecule
  Nilavo = []
  Nilavo.append([Lipinski.NumSaturatedCarbocycles(i) for i in mol])
  descriptors['Saturated_Carbocycles'] = Nilavo[0]

  # Number of saturated heterocycles for a molecule
  Nilavo = []
  Nilavo.append([Lipinski.NumSaturatedHeterocycles(i) for i in mol])
  descriptors['Saturated_Heterocycles'] = Nilavo[0]

  # Number of saturated rings for a molecule
  Nilavo = []
  Nilavo.append([Lipinski.NumSaturatedRings(i) for i in mol])
  descriptors['Saturated_Rings'] = Nilavo[0]

  # Number of rings for a molecule
  Nilavo = []
  Nilavo.append([Lipinski.RingCount(i) for i in mol])
  descriptors['Rings'] = Nilavo[0]

  return descriptors

In [11]:
regression_datasets = ['caco2_wang', 
                       'lipophilicity_astrazeneca', 
                       'solubility_aqsoldb', 
                       'ppbr_az', 
                       'vdss_lombardo', 
                       'half_life_obach', 
                       'clearance_microsome_az',
                       'clearance_hepatocyte_az', 
                       'ld50_zhu'
                      ]

In [12]:
classification_datasets = ['hia_hou', 
                       'pgp_broccatelli', 
                       'bioavailability_ma', 
                       'bbb_martins', 
                       'cyp2d6_veith', 
                       'cyp3a4_veith', 
                       'cyp2c9_veith',
                       'cyp2d6_substrate_carbonmangels', 
                       'cyp3a4_substrate_carbonmangels',
                       'cyp2c9_substrate_carbonmangels',
                       'herg',
                       'ames',
                       'dili'
                       ]

# Calculate performence of my method

* TEST ON INDIPENDENT TEST SET (which is not used in model selection)

For robust measurement of model performance, I did five independent runs of the model to calculate average performance and standard deviation.

In [13]:
from tdc.benchmark_group import admet_group
group = admet_group(path = 'data/')
predictions_list = []

for seed in [1, 2, 3, 4, 5]:
  predictions = {}

  for benchmark in group:
    name = benchmark['name']
    train_val, test = benchmark['train_val'], benchmark['test']

    # split the train_val set into train & validation set
    train, valid = group.get_train_valid_split(benchmark = name, split_type = 'scaffold', seed = seed)

    # extract features from molecule SMILES string
    x_train = molecular_descriptors(train)
    x_valid = molecular_descriptors(valid)
    x_test = molecular_descriptors(test)

    # target column
    y_train = train.Y
    y_valid = valid.Y
    y_test = test.Y

    # merging traning and validation set
    x_train_val = pd.concat([x_train, x_valid])
    y_train_val = pd.concat([y_train, y_valid], axis=0)

    # Replace NaN values with 0
    x_train_val = np.nan_to_num(x_train_val, nan=0, posinf=0)
    x_test = np.nan_to_num(x_test, nan=0, posinf=0)

    # take appropriate model for different datasets
    if name in regression_datasets:
      model = best_model_series[name]
    elif name in classification_datasets:
      model = best_Cmodel_series[name]

    # train model on training and validation set
    # and
    # test on test set 
    model.fit(x_train_val, y_train_val)
    y_pred_test = model.predict(x_test)

    predictions[name] = y_pred_test

  predictions_list.append(predictions)

results = group.evaluate_many(predictions_list)

Found local copy...
--- caco2_wang ---
generating training, validation splits...
100%|██████████| 728/728 [00:00<00:00, 2152.03it/s]
--- hia_hou ---
generating training, validation splits...
100%|██████████| 461/461 [00:00<00:00, 2828.82it/s]
--- pgp_broccatelli ---
generating training, validation splits...
100%|██████████| 973/973 [00:00<00:00, 2429.42it/s]
--- bioavailability_ma ---
generating training, validation splits...
100%|██████████| 512/512 [00:00<00:00, 2907.56it/s]
--- lipophilicity_astrazeneca ---
generating training, validation splits...
100%|██████████| 3360/3360 [00:01<00:00, 2452.55it/s]
--- solubility_aqsoldb ---
generating training, validation splits...
100%|██████████| 7985/7985 [00:01<00:00, 4670.62it/s]
--- bbb_martins ---
generating training, validation splits...
100%|██████████| 1624/1624 [00:00<00:00, 2694.55it/s]
--- ppbr_az ---
generating training, validation splits...
100%|██████████| 2231/2231 [00:00<00:00, 2323.64it/s]
--- vdss_lombardo ---
generating trai

In [14]:
results

{'ames': [0.716, 0.0],
 'bbb_martins': [0.811, 0.013],
 'bioavailability_ma': [0.523, 0.011],
 'caco2_wang': [0.321, 0.005],
 'clearance_hepatocyte_az': [0.44, 0.003],
 'clearance_microsome_az': [0.518, 0.005],
 'cyp2c9_substrate_carbonmangels': [0.281, 0.0],
 'cyp2c9_veith': [0.556, 0.0],
 'cyp2d6_substrate_carbonmangels': [0.478, 0.018],
 'cyp2d6_veith': [0.358, 0.0],
 'cyp3a4_substrate_carbonmangels': [0.605, 0.0],
 'cyp3a4_veith': [0.654, 0.0],
 'dili': [0.7, 0.0],
 'half_life_obach': [0.438, 0.011],
 'herg': [0.715, 0.011],
 'hia_hou': [0.818, 0.01],
 'ld50_zhu': [0.636, 0.001],
 'lipophilicity_astrazeneca': [0.617, 0.003],
 'pgp_broccatelli': [0.818, 0.0],
 'ppbr_az': [9.185, 0.0],
 'solubility_aqsoldb': [0.828, 0.002],
 'vdss_lombardo': [0.627, 0.01]}

| **Dataset**                    | Model                     |Matric      | Performance   | Rank |
| ------------------------------ | ------------------------- |:----------:|:-------------:|:----:|
| *`ABSORPTION`*                                                                                 |
| caco2_wang                     | AdaBoostRegressor         | MAE ↓      | 0.321 ± 0.005 | 2    |
| hia_hou                        | BaggingClassifier         | AUROC ↑    | 0.818 ± 0.01  | 9    |
| pgp_broccatelli                | ExtraTreesClassifier      | AUROC ↑    | 0.818 ± 0.0   | -    |
| bioavailability_ma             | RandomForestClassifier    | AUROC ↑    | 0.523 ± 0.011 | -    |
| lipophilicity_astrazeneca      | AdaBoostRegressor         | MAE ↓      | 0.617 ± 0.003 | 8    |
| solubility_aqsoldb             | AdaBoostRegressor         | MAE ↓      | 0.828 ± 0.002 | 4    |
| *`DISTRIBUTION`*                                                                               |
| bbb_martins                    | BaggingClassifier         | AUROC ↑    | 0.811 ± 0.013 | 10   |
| ppbr_az                        | GradientBoostingRegressor | MAE ↓      | 9.185 ± 0.0   | 2    |
| vdss_lombardo                  | AdaBoostRegressor         | SPEARMAN ↑ | 0.627 ± 0.01  | 1    |
| *`METABOLISM`*                                                                                 |
| cyp2d6_veith                   | ExtraTreeClassifier       | AUPRC ↑    | 0.358 ± 0.0   | -    |
| cyp3a4_veith                   | XGBClassifier             | AUPRC ↑    | 0.654 ± 0.0   | -    |
| cyp2c9_veith                   | BaggingClassifier         | AUPRC ↑    | 0.556 ± 0.0   | -    |
| cyp2d6_substrate_carbonmangels | BaggingClassifier         | AUPRC ↑    | 0.605 ± 0.0   | -    |
| cyp3a4_substrate_carbonmangels | XGBClassifier             | AUROC ↑    | 0.556 ± 0.0   | 7    |
| cyp2c9_substrate_carbonmangels | BaggingClassifier         | AUPRC ↑    | 0.281 ± 0.0   | -    |
| *`EXCRETION`*                                                                                  |
| half_life_obach                | AdaBoostRegressor         | SPEARMAN ↑ | 0.438 ± 0.011 | 1    |
| clearance_microsome_az         | AdaBoostRegressor         | SPEARMAN ↑ | 0.518 ± 0.005 | 8    |
| clearance_hepatocyte_az        | ExtraTreesRegressor       | SPEARMAN ↑ | 0.44 ± 0.003  | 1    |
| *`TOXICITY`*                                                                                   |
| herg                           | BaggingClassifier         | AUROC ↑    | 0.715 ± 0.011 | -    |
| ames                           | ExtraTreesClassifier      | AUROC ↑    | 0.716 ± 0.0   | -    |
| dili                           | KNeighborsClassifier      | AUROC ↑    | 0.7 ± 0.0     | -    |
| ld50_zhu                       | ExtraTreesRegressor       | MAE ↓      | 0.636 ± 0.001 | 4    |